proposals/fixed-sized-buffers.md

```diff
@@ -29,42 +31,51 @@ public fixed DXGI_RGB GammaCurve[1025];
 The declaration would get translated into an internal representation by the compiler that is similar to the following
 
 ```C#
-[CompilerGenerated, StructLayout(LayoutKind.Sequential, Pack = 1)]
-struct <GammaCurve>e__FixedBuffer
+
+[FixedBuffer(typeof(DXGI_RGB), 1024)]
+public ConsoleApp1.<Buffer>e__FixedBuffer_1024<DXGI_RGB> GammaCurve;
+
+// Pack = 0 is the default packing and should result in indexable layout.
+[CompilerGenerated, UnsafeValueType, StructLayout(LayoutKind.Sequential, Pack = 0)]
+struct <Buffer>e__FixedBuffer_1024<T>
 {
-    private DXGI_RGB _e0;
-    private DXGI_RGB _e1;
+    private T _e0;
+    private T _e1;
     // _e2 ... _e1023
-    private DXGI_RGB _e1024;
+    private T _e1024;
 
-    public ref DXGI_RGB this[int index]
-    {
-        get;
-    }
+    public ref T this[int index] => ref (uint)index <= 1024u ?
+                                         ref RefAdd<T>(ref _e0, index):
+                                         throw new IndexOutOfRange();
 }
 ```
 
+Since such fixed-sized buffers no longer require use of `fixed`, it makes sense to allow any element type.  
+
+> NOTE: `fixed` will still be supported, but only if the element type is `blittable`
```

Comment:
Today, unsafe `fixed` buffers only support a subset of types that would be considered `blittable`. Is this expected to be expanded to include all `blittable` types?

Reply:
‘blittable’ is a subset of ‘unmanaged’ types that includes types which (and their arrays) have the same native/com representation as in the managed code.  

It is a bit of a circular definition that ‘fixed’ allows only blittable types. 

If the set of blittable typea is expanded (why not enums, for example), then they will be allowed in ‘fixed’. - if that is what you are asking.

Reply:
Enums are one of the `sometimes` blittable types, as they also include `bool` and `char` (at the IL level).

`fixed` today doesn't support `IntPtr`, `UIntPtr`, or structs that are composed of only blittable types (`Vector4` or `Complex` for example).

I was wonting to confirm that after this support goes in, will those will be supported in the `unsafe` fixed-sized buffer (the one that exists today) context as well as the `safe` fixed-sized buffer (the one being proposed here) context

Reply:
I don't think we plan to make any changes to the existing unsafe fixed feature. The reason being that if we were starting from scratch today it's likely a feature we simply wouldn't add. Instead we'd focus on the experience being desigend here: safe, fixed buffers that work for all types.

Reply:
That is prpbably fine, as long as the new functionality can still be used with unsafe and interop code. And as long as you can have an unsafe struct that also contains a "safe" fixed sized buffer...

Reply:
FWIW. Once there is an indexer, you will likely be able to pin it. 
Any other ref-returning indexer can be pinned, assuming it has an unmanaged type, so it feels like indexable `fixed` should be pinnable as well.

```C#
fixed (S* p = &thing.buffer[0])
{
        // ... use p here
}
```
You can use pinned pointer for anything, including interop. That would be unsafe - with all the issues regarding marshaling, block-copying and blittability (https://en.wikipedia.org/wiki/Blittable_types.)

I think this will allow you to get your scenario working. It would be unsafe and possibly platform specific.

As for 

```C#
fixed (S* p = thing.buffer)
{
        // ... use p here
}
```

We should at least support what worked before because of compat - assuming we use the same or similar syntax. Not sure if we need to go further.

Reply:
You should also continue to be able to take the address of 'thing' (assuming 'thing' is a struct, 'thing.buffer' is a fixed array for a blittable type and the other members of 'thing' are also blittable).

I think the backing data structure being generic will hinder this scenario, unless the compiler relaxes the restriction here (taking the address of a generic type, provided it is blittable, is part of the blittable proposal).